File: 01191. K-Concatenation Maximum Sum/solution.js
/**
 * Problem: 1191. K-Concatenation Maximum Sum
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 5 ms (Beats 100%)
 */

/**
 * Returns the max subarray sum for k concatenations of arr
 *
 * @param {number[]} arr - Input array
 * @param {number} k - Number of concatenations
 *
 * @returns {number} - Max subarray sum modulo 1e9+7
 */
const kConcatenationMaxSum = (arr, k) => {
  // Define the modulo constant for the result
  const MOD = 1000000007

  // Initialize the total sum of the array, local maximum for Kadane's algorithm, and global maximum subarray sum
  let totalSum = 0,
    localMax = 0,
    globalMax = 0

  // Iterate through the array to calculate total sum and find the maximum subarray sum using Kadane's algorithm
  arr.forEach((num) => {
    totalSum += num // Accumulate the sum of all elements

    localMax = Math.max(num, localMax + num) // Update local maximum subarray sum ending at current element
    globalMax = Math.max(localMax, globalMax) // Update global maximum subarray sum found so far
  })

  // If k is 1 or the maximum subarray sum is 0, return the result modulo MOD
  if (k === 1 || globalMax === 0) return globalMax % MOD

  // Run Kadane's algorithm again to account for the case when the array is concatenated twice
  arr.forEach((num) => {
    localMax = Math.max(num, localMax + num) // Update local maximum for the second concatenation
    globalMax = Math.max(localMax, globalMax) // Update global maximum if a larger sum is found
  })

  // If the total sum of the array is positive, add the sum of the middle (k-2) arrays to the global maximum
  if (totalSum > 0) globalMax += (k - 2) * totalSum

  // Return the final result modulo MOD
  return globalMax % MOD
}

File: 01191. K-Concatenation Maximum Sum/solution.go
/**
 * Problem: 1191. K-Concatenation Maximum Sum
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func kConcatenationMaxSum(arr []int, k int) int {
	// Initialize variables:
	// totalSum: cumulative sum of array elements
	// maxPrefixSum: maximum prefix sum found so far
	// minPrefixSum: minimum prefix sum found so far
	// maxSubarraySum: maximum subarray sum found so far
	var totalSum, maxPrefixSum, minPrefixSum, maxSubarraySum int

	// Iterate through each element in arr
	for _, num := range arr {
		// Update cumulative sum
		totalSum += num
		// Update maximum prefix sum
		maxPrefixSum = max(maxPrefixSum, totalSum)
		// Update minimum prefix sum
		minPrefixSum = min(minPrefixSum, totalSum)
		// Update maximum subarray sum using the difference between current sum and minimum prefix sum
		maxSubarraySum = max(maxSubarraySum, totalSum-minPrefixSum)
	}

	const mod = 1e9 + 7 // Define modulo constant for result

	// Initialize answer with maximum subarray sum found in single array
	ans := maxSubarraySum

	// If k == 1, only one copy of arr is used, so return answer modulo mod
	if k == 1 {
		return ans % mod
	}

	// Calculate maximum suffix sum (sum from some index to end)
	maxSuffixSum := totalSum - minPrefixSum
	// Update answer considering concatenation of two arrays
	ans = max(ans, maxSuffixSum+maxPrefixSum)

	// If totalSum > 0, consider using more than two arrays for maximum sum
	if totalSum > 0 {
		// Update answer considering k arrays: prefix + (k-2) * totalSum + suffix
		ans = max(ans, maxSuffixSum+(k-2)*totalSum+maxPrefixSum)
	}

	// Return final answer modulo mod
	return ans % mod
}
